File: 1.zig
//! https://adventofcode.com/2023/day/1

const std = @import("std");

const digits = [_][]const u8{ "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const alloc = arena.allocator();
    const args = try std.process.argsAlloc(alloc);
    const file = try std.fs.cwd().openFile(args[1], .{});
    defer file.close();
    const input = try file.readToEndAlloc(alloc, std.math.maxInt(u32));

    var parts = [2]usize{ 0, 0 };
    for (&parts, 1..) |*partsum, partnum| {
        var lines = std.mem.tokenizeScalar(u8, input, '\n');
        while (lines.next()) |line| {
            // std.debug.print("line='{s}'\n", .{line});
            const first_digit = blk: for (0..line.len) |i| {
                const c = line[i];
                if (std.ascii.isDigit(c)) break c - '0';
                if (partnum == 1) continue;
                for (digits, 1..) |dig, digi| {
                    if (std.mem.startsWith(u8, line[i..], dig))
                        break :blk @as(u8, @truncate(digi));
                }
            } else return error.NoDigtFound;
            // std.debug.print("first_digit={}\n", .{first_digit});

            var iter = std.mem.reverseIterator(line);
            const last_digit = blk: while (iter.next()) |c| {
                if (std.ascii.isDigit(c)) break :blk c - '0';
                if (partnum == 1) continue;
                for (digits, 1..) |dig, digi| {
                    if (std.mem.startsWith(u8, line[iter.index..], dig))
                        break :blk @as(u8, @truncate(digi));
                }
            } else return error.NoDigtFound;

            partsum.* += first_digit * 10 + last_digit;
        }
    }

    std.debug.print("parts={any}\n", .{parts});
}

File: 1.rs
use std::env;
use std::fs::File;
use std::io::prelude::*;

pub fn main2() -> std::io::Result<()> {

    let path = env::args().skip(1).next().unwrap();
    let mut file = File::open(path)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    let mut sum: u32 = 0;
    const DIGIT_NAMES: &'static [&'static str] = &["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
    for line in contents.split('\n') {
        if line.len() == 0 { continue; }
        let mut first_digit = None;
        'outer: for i in 0..line.len() {
            let c = line.chars().nth(i).unwrap();
            if c.is_ascii_digit() { first_digit = Some(c); break ;}
            for (j, dn) in DIGIT_NAMES.iter().enumerate() {
                // eprintln!("{j} {dn}");
                if line[i..].starts_with(dn) { 
                    first_digit = char::from_u32((j + 49) as u32); break 'outer;
                }
            }
        }
        let mut last_digit = None;
        'outer: for i in (0..line.len()).rev() {
            let c = line.chars().nth(i).unwrap();
            if c.is_ascii_digit() { last_digit = Some(c); break ;}
            for (j, dn) in DIGIT_NAMES.iter().enumerate() {    
                if line[i..].starts_with(dn) { 
                    last_digit = char::from_u32((j + 49) as u32); break 'outer; 
                }
            }
        }
        // eprintln!("{line} {first_digit:?} {last_digit:?}");
        sum += first_digit.unwrap().to_digit(10).unwrap() * 10 + last_digit.unwrap().to_digit(10).unwrap();
    }
    println!("sum {sum}");
    Ok(())
}

const STR_DIGITS: &[&[u8]] = &[b"one", b"two", b"three", b"four", b"five", b"six", b"seven", b"eight", b"nine"];

fn digit_sum(line: &[u8], p2: bool) -> usize {
    let mut digits = (0..line.len()).filter_map(|i| match line[i] {
        b'0'..=b'9' => Some((line[i] - b'0') as usize),
        _ if(p2) => STR_DIGITS.iter()
        .enumerate()
        .find_map(|(di, d)| line[i..].starts_with(d).then_some(di+1)),
        _ => None,
    });

    let a = digits.next().unwrap();
    let b = digits.last().unwrap_or(a);
    return a*10 + b;
}

fn main() -> std::io::Result<()> {
    let path = env::args().skip(1).next().unwrap();
    let mut file = File::open(path)?;
    let mut input = String::new();
    file.read_to_string(&mut input)?;
    let lines = input.trim().split('\n').map(str::as_bytes).collect::<Vec<_>>();
    let p1: usize = lines.iter().map(|line| digit_sum(line, false)).sum();
    let p2: usize = lines.iter().map(|line| digit_sum(line, true)).sum();
    eprintln!("{p1} {p2}");
    Ok(())
}